### ISSUE TITLE
FedAvg implementation does not synchronize BatchNorm running statistics (implicit FedBN behavior)

### ISSUE THREAD

**Issue by Raindow** [TsingZ0/PFL-Non-IID#260]
**Description**

I noticed a potential issue in the current FedAvg implementation regarding the handling of BatchNorm (BN) layers, specifically the running statistics (running_mean, running_var, and num_batches_tracked).

In the server-side aggregation logic, only model parameters (i.e., model.parameters()) are aggregated and redistributed to clients. However, BatchNorm running statistics are registered as buffers rather than parameters, and therefore:

They are not aggregated on the server

They are not synchronized across clients

Each client continues to use its own local BN running statistics during training and evaluation

As a result, although the code is labeled as FedAvg, the actual behavior corresponds to an implicit FedBN-style setup, where BN statistics are client-specific.

**Evidence from the code**

In the server aggregation logic:

self.global_model = copy.deepcopy(self.uploaded_models[0])
for param in self.global_model.parameters():
    param.data.zero_()

for w, client_model in zip(self.uploaded_weights, self.uploaded_models):
    self.add_parameters(w, client_model)

and:

for server_param, client_param in zip(
        self.global_model.parameters(),
        client_model.parameters()):
    server_param.data += client_param.data.clone() * w

Only parameters() are aggregated.
BatchNorm buffers (running_mean, running_var) are never aggregated or synchronized.

When models are sent back to clients, each client therefore keeps its own BN statistics, which depend on its local data distribution.

**Why this is problematic**

From a methodological perspective:

Standard FedAvg assumes that the entire model state is shared and synchronized across clients each round.

In PyTorch, this corresponds to synchronizing the full state_dict(), not only parameters().

BatchNorm running statistics are part of the model’s functional state;
identical weights with different BN statistics define different models.

Unless explicitly stated, a FedAvg implementation should not leave BN behavior client-specific.

Currently:

The method behaves like FedBN (BN kept local)

But this behavior is implicit, undocumented, and not controlled by design

This may lead to misleading comparisons, especially when evaluating FedAvg baselines

**Expected behavior**

There are two valid and clearly distinguishable design choices:

True FedAvg behavior

Aggregate and distribute the full state_dict()

BN running statistics are synchronized across clients

Explicit FedBN-style behavior

Keep BN statistics local by design

Clearly document this choice

Optionally exclude BN buffers from aggregation explicitly

However, the current implementation does neither explicitly:

It does not synchronize BN statistics

But also does not document or justify BN localization

**Suggested fix**

If the intention is standard FedAvg, the server should aggregate and distribute the full model state:

global_model.load_state_dict(aggregated_state_dict)

If the intention is FedBN, then BN layers should be explicitly excluded and clearly documented as client-specific.

**Summary**

The current FedAvg implementation does not synchronize BatchNorm running statistics

This results in implicit FedBN behavior

This should be either:

fixed (by synchronizing full state_dict()), or

explicitly documented and controlled as a design choice

Clarifying this will improve correctness, reproducibility, and fairness in experimental comparisons.

**Comment by TsingZ0** [comment]
see #84
